### PR TITLE
Port 9402 port api search identifier

### PR DESCRIPTION
**Pull request by talsabagport**
# Description

What - Add support for search identifier feature
Why - Allows you to update entity without knowing its identifier
How - Refactor the code to allow run initial sync independently, which then can be used to evaluate and keep the identifiers while running initial sync to Port. The identifiers needed for the delete stale entities process at the end of the initial resync.

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

### CODE REVIEW

File: pkg/port/entity/entity.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
+package entity
```
Comment (danielsinai) [inline pkg/port/entity/entity.go:1]: Why we created a new entity package?
Shouldn't it be under cli?

File: pkg/port/entity/entity.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+func newEntityRequest(obj interface{}, mapping port.EntityMapping) (*port.EntityRequest, error) {
+	var err error
+	entity := &port.EntityRequest{}
+
+	if reflect.TypeOf(mapping.Identifier).Kind() == reflect.String {
+		entity.Identifier, err = jq.ParseString(mapping.Identifier.(string), obj)
+	} else if reflect.TypeOf(mapping.Identifier).Kind() == reflect.Map {
+		entity.Identifier, err = jq.ParseMapRecursively(mapping.Identifier.(map[string]interface{}), obj)
+	}
```
Comment (danielsinai) [inline pkg/port/entity/entity.go:90]: If none of the above will catch (for example an array / number), we will get an unexpected error, maybe worth adding else and write an error about this type not being support

File: pkg/handlers/controllers_test.go [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+			if err != nil {
+				f.t.Errorf("error deleting object %s: %v", d.name, err)
+			}
+		}
+	}
+}
+
+func (f *fixture) assertObjectsHandled(objects []struct{ kind, name string }) {
+	assert.Eventually(f.t, func() bool {
+		integrationKinds, err := f.portClient.GetIntegrationKinds(f.controllersHandler.stateKey)
```
Comment (danielsinai) [inline pkg/handlers/controllers_test.go:380]: Nice :)

File: pkg/jq/parser.go [hunk 4]
```diff
@@ -117,17 +117,23 @@ func ParseMapInterface(jqQueries map[string]string, obj interface{}) (map[string
...
-func ParseRelations(jqQueries map[string]interface{}, obj interface{}) (map[string]interface{}, error) {
+func ParseMapRecursively(jqQueries map[string]interface{}, obj interface{}) (map[string]interface{}, error) {
 	mapInterface := make(map[string]interface{}, len(jqQueries))
 
 	for key, jqQuery := range jqQueries {
 
 		if reflect.TypeOf(jqQuery).Kind() == reflect.String {
-			queryRes, _ := ParseMapInterface(map[string]string{key: jqQuery.(string)}, obj)
+			queryRes, err := ParseMapInterface(map[string]string{key: jqQuery.(string)}, obj)
+			if err != nil {
```
Comment (danielsinai) [inline pkg/jq/parser.go:127]: Nice catch!

File: pkg/jq/parser.go [hunk 5]
```diff
@@ -139,7 +145,7 @@ func ParseRelations(jqQueries map[string]interface{}, obj interface{}) (map[stri
 			jqArrayValue := reflect.ValueOf(jqQuery)
 			relations := make([]interface{}, jqArrayValue.Len())
 			for i := 0; i < jqArrayValue.Len(); i++ {
-				relation, err := ParseRelations(map[string]interface{}{key: jqArrayValue.Index(i).Interface()}, obj)
+				relation, err := ParseMapRecursively(map[string]interface{}{key: jqArrayValue.Index(i).Interface()}, obj)
 				if err != nil {
 					return nil, err
 				}
```
Comment (danielsinai) [inline pkg/jq/parser.go:151]: May you also add here the else?

**Review by danielsinai — CHANGES_REQUESTED** [review]
Great work 🤩

File: pkg/jq/parser_test.go
Comment (omby8888) [inline pkg/jq/parser_test.go:1]: add a test on search identifier

**Review by omby8888 — APPROVED** [review]
LGTM

**Review by danielsinai — APPROVED** [review]

**Pull request merged**